A15-binary-logging.md

```diff
@@ -0,0 +1,352 @@
+    EVENT_TYPE_CLIENT_HEADER = 1;
+    // Headers sent from server to client
+    EVENT_TYPE_SERVER_HEADER = 2;
+    // Message sent from client to server
+    EVENT_TYPE_CLIENT_MESSAGE = 3;
+    // message sent from server to client
+    EVENT_TYPE_SERVER_MESSAGE = 4;
+    // Signal that client is done sending
+    EVENT_TYPE_CLIENT_HALF_CLOSE = 5;
+    // Trailers sent from server to client.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_SERVER_TRAILER = 6;
+    // Signal that the RPC is cancelled.
+    // This marks the end of the RPC. Events may arrive after
+    // this due to races. The extra events are not considered
+    // a part of the RPC and may be safely ignored.
+    EVENT_TYPE_CANCEL = 7;
+  }
+
+  // Enumerates the entity that generates the log entry
+  enum Logger {
+    LOGGER_UNKNOWN = 0;
+    LOGGER_CLIENT = 1;
+    LOGGER_SERVER = 2;
+  }
+
+  EventType type = 1;
+  Logger logger = 2;  // One of the above Logger enum
+
+  // Uniquely identifies a call. Each call may have several log entries, they
+  // will all have the same call_id. Nothing is guaranteed about their values
+  // other than they are unique across different RPCs in the same gRPC process.
+  int64 call_id = 3;
+
+  // The logger uses one of the following fields to record the payload,
+  // according to the type of the log entry.
+  oneof payload {
+    // Used by EVENT_TYPE_CLIENT_HEADER, EVENT_TYPE_SERVER_HEADER and
+    // EVENT_TYPE_SERVER_TRAILER. This contains only the metadata
+    // from the application.
+    Metadata metadata = 4;
+    // Used by EVENT_TYPE_CLIENT_MESSAGE, EVENT_TYPE_SERVER_MESSAGE
+    Message message = 5;
+  }
+
+  // Peer address information, will only be recorded in
+  // EVENT_TYPE_CLIENT_HEADER or EVENT_TYPE_SERVER_HEADER
+  Peer peer = 6;
+
+  // true if payload does not represent the full message or metadata.
+  bool truncated = 7;
+
+  // The method name. Logged for EVENT_TYPE_CLIENT_HEADER
+  string method_name = 8;
+
+  // status_code and status_message:
+  // Only present for EVENT_TYPE_SERVER_TRAILER.
+  uint32 status_code = 9;
+
+  // An original status message before any transport specific
+  // encoding.
+  string status_message = 10;
+
+  // The value of the 'grpc-status-details-bin' metadata key. If
+  // present, this is always an encoded 'google.rpc.Status' message.
+  bytes status_details = 11;
+
+  // the RPC timeout
+  google.protobuf.Duration timeout = 12;
+
+  // The entry sequence id for this call. The first GrpcLogEntry has a
+  // value of 1, to disambiguate from an unset value. The purpose of
+  // this field is to detect missing entries in environments where
+  // durability or ordering is not guaranteed.
+  uint32 sequence_id_within_call = 13;
+
+  // The timestamp of the binary log message
+  google.protobuf.Timestamp timestamp = 14;
```

Comment:
My 2c: timestamp should be the first field.  Actually, a full reordering might be worth it.  Something like this seems most logical to me:

- timestamp, call ID, sequence ID
- logger, event type
- payload (metadata / message), truncated
- header fields: method name, peer, timeout
- trailer fields: status code, message, details

Also, we may want sub-messages for header/trailer fields.

Reply:
+1 on the re-ordering comment.

I'm ambivalent about sub-messages for header vs. trailer fields.

Reply:
I reordered the fields and grouped header and trailer into sub messages. It ended up looking pretty reasonable so I support grouping them.

I was considering making `peer` always be logged at SERVER_HEADER because on client side CLIENT_HEADER is before name resolution / load balancing, and on server side we can easily defer the logging until SERVER_HEADER. However, in the case where the RPC gets cancelled before SERVER_HEADER, the server would not be able to log the peer info.

I'm going to leave it as a top level field that appears at the SERVER_HEADER event on client side and CLIENT_HEADER for server side.